types/shares.go

```diff
@@ -56,35 +56,107 @@ func (m Message) MarshalDelimited() ([]byte, error) {
+// appendToShares appends raw data as shares.
+// Used for messages.
+func appendToShares(shares []NamespacedShare, nid namespace.ID, rawData []byte) []NamespacedShare {
+	const adjustedSize = ShareSize - NamespaceSize
+	if len(rawData) < adjustedSize {
+		rawShare := []byte(append(nid, rawData...))
+		paddedShare := zeroPadIfNecessary(rawShare, ShareSize)
+		share := NamespacedShare{paddedShare, nid}
+		shares = append(shares, share)
+	} else { // len(rawData) >= adjustedSize
+		shares = append(shares, split(rawData, nid)...)
+	}
+	return shares
+}
+
+// splitContiguous splits multiple raw data contiguously as shares.
+// Used for transactions, intermediate state roots, and evidence.
+func splitContiguous(nid namespace.ID, rawDatas [][]byte) []NamespacedShare {
+	shares := make([]NamespacedShare, 0)
+	const adjustedSize = ShareSize - NamespaceSize - ShareReservedBytes
+	// Index into the outer slice of rawDatas
+	outerIndex := 0
+	// Index into the inner slice of rawDatas
+	innerIndex := 0
+	for outerIndex < len(rawDatas) {
+		var rawData []byte
+		startIndex := 0
+		rawData, outerIndex, innerIndex, startIndex = getNextChunk(rawDatas, outerIndex, innerIndex, adjustedSize)
+		rawShare := []byte(append(append(nid, byte(startIndex)), rawData...))
+		paddedShare := zeroPadIfNecessary(rawShare, ShareSize)
 		share := NamespacedShare{paddedShare, nid}
 		shares = append(shares, share)
-	} else { // len(rawData) >= shareSize
-		shares = append(shares, split(rawData, shareSize, nid)...)
 	}
 	return shares
 }
 
 // TODO(ismail): implement corresponding merge method for clients requesting
 // shares for a particular namespace
-func split(rawData []byte, shareSize int, nid namespace.ID) []NamespacedShare {
+func split(rawData []byte, nid namespace.ID) []NamespacedShare {
+	const adjustedSize = ShareSize - NamespaceSize
 	shares := make([]NamespacedShare, 0)
-	firstRawShare := rawData[:shareSize]
+	firstRawShare := []byte(append(nid, rawData[:adjustedSize]...))
 	shares = append(shares, NamespacedShare{firstRawShare, nid})
-	rawData = rawData[shareSize:]
+	rawData = rawData[adjustedSize:]
 	for len(rawData) > 0 {
-		shareSizeOrLen := min(shareSize, len(rawData))
-		paddedShare := zeroPadIfNecessary(rawData[:shareSizeOrLen], shareSize)
+		shareSizeOrLen := min(adjustedSize, len(rawData))
+		rawShare := []byte(append(nid, rawData[:shareSizeOrLen]...))
+		paddedShare := zeroPadIfNecessary(rawShare, ShareSize)
 		share := NamespacedShare{paddedShare, nid}
 		shares = append(shares, share)
 		rawData = rawData[shareSizeOrLen:]
 	}
 	return shares
 }
 
+func getNextChunk(rawDatas [][]byte, outerIndex int, innerIndex int, width int) ([]byte, int, int, int) {
+	rawData := make([]byte, 0, width)
+	startIndex := 0
+	firstBytesToFetch := 0
+
+	curIndex := 0
+	for curIndex < width && outerIndex < len(rawDatas) {
+		bytesToFetch := min(len(rawDatas[outerIndex])-innerIndex, width-curIndex)
+		// Prune any extra 0-data (this should never happen, but just in case)
```

Comment:
"this should never happen" in the sense of "this is an invalid state" kind of way?

Reply:
Yes. The only way to get in these is with 0-length transactions/evidence/ISRs, which should be invalid. Should this be a panic instead?

Reply:
> Should this be a panic instead?

Yeah, that's why I was asking. Because otherwise, we are reaching an invalid / unexpected state and silently recovering from it without even logging it. But we have to be careful here: is there any chance a (malicious) node could misuse panicing to shut down another honest node?

Reply:
The mempool logic should reject 0-length txs and evidence, and ISRs are non-zero by definition. I don't know if it does that currently.

Reply:
Not sure either. Let's panic on this and we'll find out soon enough :-D Nah, seriously, if this is some state that should never be reached (and if panicing can't be misused - or we can make sure it can't be misused), we should panic. But there should be an extensive comment explaining the expectations.

Reply:
Should the mempool logic be incorporated into ll-core or to the app via `CheckTx`? Both?

Reply:
I feel like such a fundamental check could just be part of ll-core. There's zero way a 0-length tx or evidence is valid.

Reply:
Fixed in ff636c4. Should we open an issue to make sure ll-core doesn't allow 0-length requests?

Reply:
Yes, probably a good idea to tack this in an issue.